torch/_inductor/codegen/simd.py

```diff
@@ -1748,9 +1748,12 @@ def _bench(candidate_split_size):
                 partial_accum.reduction_type, partial_accum.reduction_type
             )
 
-            V.graph.wrapper_code.writeline(
-                f"{buffer_name} = {ws_name}[{start} : {end}].view({nsplit}, {rnumel}).{opname}(dim=0)",
-            )
+            final_reduce = f"{buffer_name} = {ws_name}[{start} : {end}].view({nsplit}, {rnumel}).{opname}(dim=0)"
+            # The workspace tensor is in torch.float, need a cast if the buffer is
+            # not.
+            if (buffer_dtype := V.graph.get_dtype(buffer_name)) != torch.float:
```

Comment:
I don't think it's strictly `float`. its more that its low precision and accum is done in fp32. i did a pr for this for split reductions, let me try to find

Reply:
https://github.com/pytorch/pytorch/blob/171ae4a8d20781d2853126f6913accfa291f7e30/torch/_inductor/ir.py#L1897-L1903

but i think its just low precision dtype float dtype -> fp32

Reply:
We allocate the workspace tensor in fp32. So if the buffer has a different dtype, we should cast.

Is your concern about even higher precision like fp64 inputs?

Reply:
I see. For fp64, we should allocate the workspace buffer in fp64 in the first place

Reply:
but for f16 as well, we shouldnt write to it in fp32, or it will truncate on store.